rspirv/dr/build/mod.rs

```diff
@@ -90,23 +90,92 @@ type BuildResult<T> = result::Result<T, Error>;
 pub struct Builder {
     module: dr::Module,
     next_id: u32,
-    function: Option<dr::Function>,
-    block: Option<dr::Block>,
+    new_function: Option<dr::Function>,
+    new_block: Option<dr::Block>,
+    selected_function: Option<usize>,
+    selected_block: Option<usize>,
     version: Option<(u8, u8)>,
 }
 
+pub enum InsertPoint {
+    Begin,
+    End,
+    FromBegin(usize),
+    FromEnd(usize),
+}
+
 impl Builder {
     /// Creates a new empty builder.
     pub fn new() -> Builder {
         Builder {
             module: dr::Module::new(),
             next_id: 1,
-            function: None,
-            block: None,
+            new_function: None,
+            selected_function: None,
+            new_block: None,
+            selected_block: None,
             version: None,
         }
     }
 
+    /// Create a new builder from an existing module
+    pub fn new_from_module(module: dr::Module) -> Builder {
+        let next_id = module.header.as_ref().map(|h| h.bound).expect("Expecting ModuleHeader with valid bound");
+        let version = module.header.as_ref().map(|h| h.version());
+
+        Builder {
+            module,
+            next_id,
+            new_function: None,
+            selected_function: None,
+            new_block: None,
+            selected_block: None,
+            version
+        }
+    }
+
+    fn insert_into_block(&mut self, insert_point: InsertPoint, inst: dr::Instruction) -> BuildResult<()> {
+        let allow = self.new_block.is_some() || self.selected_function.is_some() && self.selected_block.is_some();
+
+        if !allow {
+            return Err(Error::DetachedInstruction);
+        }
+
+        let ref mut block = if self.new_block.is_some() {
+            self.new_block.as_mut().unwrap()
+        } else {
+            &mut self.module.functions[self.selected_function.unwrap()].blocks[self.selected_block.unwrap()]
+        };
+
+        match insert_point {
+            InsertPoint::End => block.instructions.push(inst),
+            InsertPoint::Begin => block.instructions.insert(0, inst),
```

Comment:
This is also concerning: an application that adds N instructions to a block of M will spend O(N*M) time moving bytes around. I don't think this is acceptable to DR level of abstraction.

Reply:
Since builder is the most ergonomic way of building / patching SPIR-V that we have this kind of functionality is needed. One other way that's kind of iffy is to be able to insert bb's in arbitrary places, or just change the underlaying datatype to something more amenable to inserts in arbitrary places. I didn't change the datatype because I'd rather have the "new" code be slow then to perf-break existing code in this case.

Reply:
I see, that makes sense.
Would it be good to record the list of inserted instructions on the side, so that we can merge them in a single pass at the end of building, instead of moving bytes around on every insertion?

Reply:
I think having a delayed insertion scheme will be much more bug prone; it's probably easier to replace the Vec with a VecDeque and comment that the relative insertion points incur a performance overhead.

Reply:
I guess we can follow up with this without braking the API, so it's not a blocker anyway